neysofu/ferrum-fix
# Compile against supported openssl version

## Purpose
The crate [`openssl`](https://docs.rs/openssl/0.10.32/openssl/index.html) has a struct called [`SslOptions`](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslOptions.html) that provides __bit flags__.
These flags can be used to controll the behavior of an [`SslContext`](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslContext.html).

> Bit Flags (also: Bit Fields) can be used to store more than one  
boolean value in one byte or a whole set of bytes.

The use of one particular bit flag is currently causing a compile   
error on machines running Ubuntu 18.04 LTS, that tries to build  
FerrumFIX without having OpenSSL version 1.1.1 or newer installed. 

This pull request removes the build error on Ubuntu 18.04 LTS 
by adding conditional compilation to the lines of code that are causing the error.

### Problem
In file [`fefix/src/fixs/mod.rs`](https://github.com/neysofu/ferrum-fix/blob/f890896c18ec91f8d5cae8ab08bef4e098439ad7/fefix/src/fixs/mod.rs#L91) the SslOptions bit flag [`NO_TLSV1_3`](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslOptions.html#associatedconstant.NO_TLSV1_3)  
is used to disable the use of TLSv1.3.

However, on Ubuntu 18.04 LTS, the use of bit flag `NO_TLSV1_3` 
introduces a dependency on the target machine - requiring OpenSSL 1.1.1 
or newer to be installed. If OpenSSL 1.1.1 or newer is not installed, then the 
Rust compiler will __fail__  with the  __error message__:

```bash
error[E0599]: no associated item named `NO_TLSV1_3` found for struct `openssl::ssl::SslOptions` in the current scope
   --> fefix/src/fixs/mod.rs:118:49
    |
118 |                 context.set_options(SslOptions::NO_TLSV1_3);
    |                                                 ^^^^^^^^^^
    |                                                 |
    |                                                 associated item not found in `openssl::ssl::SslOptions`
    |                                                 help: there is an associated constant with a similar name: `NO_TLSV1_1`

error: aborting due to previous error
```


## Approach
This Pull Request provides an easy solution to the described problem  
by using conditional compilation, as proposed by @neysofu .  

By adding a _configure attribute_ to the `NO_TLSV1_3` bit flag, the Rust  
compiler will ignore the bit flag unless OpenSSL 1.1.1 or newer is  
installed on the machine at compile time.

```rust
// In file: fefix/src/fixs/mod.rs
#[cfg(openssl111)]
context.set_options(SslOptions::NO_TLSV1_3);
```
### Tested solution
> Machine used: Dell Latitude E7470, running Ubuntu 18.04LTS.

The solution has been tested by the following procedure:
1. Fork the latest ferrum-fix commit on branch `develop` (commit: [dd03638](https://github.com/neysofu/ferrum-fix/commit/dd036389252ed932b3f88e59ff258541563dc80b) )
2. Git clone and run `cargo build` locally
3. Verify that `cargo build` fails due to openssl dependency error
4. Make changes in a seperate branch
5. Run `cargo build` and verify that openssl error disapears

#### Open Questions and Pre-Merge TODOs
This pull request was only tested on Ubuntu version 18.04 LTS.
Some open (and potentially high priority) questions  remain:
- [ ] Does the same `cargo build` error occur on machines running LibreSSL?
- [ ] If the problem would turn out to exists on machines running LibreSSL - does the 
solution proposed in this pull request solve the issue? 
- [ ] Is TLS 1.3 supported in LibreSSL, and if so, from which version and onward?
- [ ] Does the same problem exist for machines running Windows?
- [ ] Does support for TLS 1.3 varry between clients and servers? 


## Learning
Preparation for this pull request mostly consisted of reading about 
conditional compilation and getting acquainted with [openssl](https://docs.rs/openssl/0.10.32/openssl/index.html).

### Sources used
+   [Conditional compilation](https://doc.rust-lang.org/stable/reference/conditional-compilation.html) 
+   Crate [openssl](https://docs.rs/openssl/0.10.32/openssl/index.html)
+   Bit Flag [`NO_TLSV1_3](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslOptions.html#associatedconstant.NO_TLSV1_3)
+   LibreSSL 3.1.1 [release notes](https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-3.1.1-relnotes.txt)
